### PR TITLE
swagger: example to document a legacy endpoint

### PR DESCRIPTION
**Pull request by COil**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | yes/no
| Tickets       | NA.
| License       | MIT
| Doc PR        | todo once this PR is merged.

This is a simple example showing how to extend the OpenAPI spec is order to document a legacy endpoint.

### CODE REVIEW

File: api/config/services.yaml [hunk 1]
```diff
@@ -41,3 +41,9 @@ services:
         decorates: App\Repository\TopBook\TopBookDataRepository
 
     App\Repository\TopBook\TopBookDataInterface: '@App\Repository\TopBook\TopBookDataRepository'
+
+    # Swagger doc customisation
+    App\Swagger\SwaggerDecorator:
+        decorates: 'api_platform.swagger.normalizer.api_gateway'
+        arguments: ['@api_platform.swagger.normalizer.api_gateway.inner' ]
```
Comment (alanpoulain) [inline api/config/services.yaml:48]: The inner argument should not be necessary with the autowiring.

File: api/src/Controller/LegacyApiController.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+declare(strict_types=1);
+
+namespace App\Controller;
+
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+
+class LegacyApiController extends AbstractController
```
Comment (alanpoulain) [inline api/src/Controller/LegacyApiController.php:11]:
```suggestion
final class LegacyApiController extends AbstractController
```

File: README.md [hunk 3]
```diff
@@ -22,25 +22,35 @@ What can I find in this demo?
 
 This demo application contains several things you may be interested.   
 
-Tests
------
+API testing
```
Comment (alanpoulain) [inline README.md:25]:
```suggestion
API Testing
```

File: README.md [hunk 4]
```diff
@@ -22,25 +22,35 @@ What can I find in this demo?
...
 
 * [Data providers documentation](https://api-platform.com/docs/core/data-providers/)
 * [Code in api/src/DataProvider](api/src/DataProvider)
 
+Overriding the OpenAPI Specification
+------------------------------------
+
+This example shows how to document an API endpoint that isn't handled by API Platform.
+This "legacy" endpoint is listed and testable like the other ones thanks to the
+Swagger interface.
```
Comment (alanpoulain) [inline README.md:49]:
```suggestion
Swagger interface.
```

File: api/config/services.yaml (hunk 1 above)
Comment (COil) [reply api/config/services.yaml:48]: Thanks @alanpoulain , yes indeed. `The autoconfigure: false` option seems useless too.

File: api/config/services.yaml [hunk 5]
```diff
@@ -41,3 +41,9 @@ services:
         decorates: App\Repository\TopBook\TopBookDataRepository
 
     App\Repository\TopBook\TopBookDataInterface: '@App\Repository\TopBook\TopBookDataRepository'
+
+    # Swagger doc customisation
+    App\Swagger\SwaggerDecorator:
+        decorates: 'api_platform.swagger.normalizer.api_gateway'
+        #arguments: ['@api_platform.swagger.normalizer.api_gateway.inner' ]
+        #autoconfigure: false
```
Comment (alanpoulain) [inline api/config/services.yaml]: Maybe the `autoconfigure` is useful here. You don't want to use this normalizer for other normalizations.

Comment (COil) [reply api/config/services.yaml]: Ok, fixed. Thanks for the review @alanpoulain :)

File: api/src/Controller/LegacyApiController.php [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+
+final class LegacyApiController extends AbstractController
+{
+    /**
+     * @Route("/stats", name="api_book_stats")
+     */
+    public function stats(): Response
```
Comment (vincentchalamon) [inline api/src/Controller/LegacyApiController.php:16]:
```suggestion
    public function __invoke(): Response
```

File: api/src/Controller/LegacyApiController.php [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+declare(strict_types=1);
+
+namespace App\Controller;
+
+use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\Routing\Annotation\Route;
+
+final class LegacyApiController extends AbstractController
```
Comment (vincentchalamon) [inline api/src/Controller/LegacyApiController.php:11]: Please add a comment explaining what is this controller for

File: api/src/Swagger/SwaggerDecorator.php [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                ],
+            ],
+        ];
+
+        $docs['paths']['/stats']['get'] = $statsEnpoint;
+
+        return $docs;
+    }
+
+    public function supportsNormalization($data, string $format = null): bool
```
Comment (vincentchalamon) [inline api/src/Swagger/SwaggerDecorator.php:57]:
```suggestion
    /**
     * {@inheritdoc}
     */
    public function supportsNormalization($data, string $format = null): bool
```

**Review by vincentchalamon — APPROVED** [review]

**Review by vincentchalamon — APPROVED** [review]

**Comment by vincentchalamon** [comment]
Thanks @COil

**Pull request merged**